[torchelastic] ensure grandchild processes are restarted correctly

zdevito opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #113481
* __->__ #113231

When torchelastic notices that one rank has failed, it will sent a SIGTERM
signal to other trainer ranks to tear them down before restarting. However,
if the trainer itself launches subprocesses, or is launched by a non-python
wrapper script, then the SIGTERM will be delivered only to the direct child of
torch eleastic and not all descendants. This opens subprocesses in a new
linux 'session' which starts a new process group with the pgid the same
as the trainers pid. Then when we send signals, we deliver them to the
process group rather than just the direct child.

zdevito committed:
[torchelastic] ensure grandchild processes are restarted correctly

When torchelastic notices that one rank has failed, it will sent a SIGTERM
signal to other trainer ranks to tear them down before restarting. However,
if the trainer itself launches subprocesses, or is launched by a non-python
wrapper script, then the SIGTERM will be delived only to the direct child of
torch eleastic and not all descendants. This opens subprocesses in a new
linux 'session' which starts a new process group with the pgid the same
as the trainers pid. Then when we send signals, we deliver them to the
process group rather than just the direct child.

[ghstack-poisoned]

H-Huang commented on torch/distributed/elastic/multiprocessing/api.py:
```diff
@@ -575,6 +575,9 @@ def __init__(
     def _popen(self, args: Tuple, env: Dict[str, str]) -> subprocess.Popen:
+        kwargs = {}
+        if not IS_WINDOWS:
+            kwargs['start_new_session'] = True
```
if new processes are started this way, they they be detached from the parent? For example, if I run something like `torchrun my_script.py  --nnodes=1 --nproc-per-node=2` and then I Ctrl+C, will the child processes still be able to exit?

zdevito replied:
The will be in a different session than the parent, but `torchrun` is a kind of like bash in that it is a process manager, so when it gets a SIGINT, it will propagate that to its children before existing. So when you Ctrl-C torchrun you get:

```
[2023-11-09 10:53:46,286] torch.distributed.elastic.agent.server.api: [WARNING] Received Signals.SIGINT death signal, shutting down workers
```

And the workers will stop too.

H-Huang replied:
Sounds good, thanks for confirming!

H-Huang reviewed (commented):
cc @kiukchung @d4l3k if you have feedback as well

zdevito committed:
Update on "[torchelastic] ensure grandchild processes are restarted correctly"


When torchelastic notices that one rank has failed, it will sent a SIGTERM
signal to other trainer ranks to tear them down before restarting. However,
if the trainer itself launches subprocesses, or is launched by a non-python
wrapper script, then the SIGTERM will be delivered only to the direct child of
torch eleastic and not all descendants. This opens subprocesses in a new
linux 'session' which starts a new process group with the pgid the same
as the trainers pid. Then when we send signals, we deliver them to the
process group rather than just the direct child.

[ghstack-poisoned]

zdevito committed:
Update on "[torchelastic] ensure grandchild processes are restarted correctly"


When torchelastic notices that one rank has failed, it will sent a SIGTERM
signal to other trainer ranks to tear them down before restarting. However,
if the trainer itself launches subprocesses, or is launched by a non-python
wrapper script, then the SIGTERM will be delivered only to the direct child of
torch eleastic and not all descendants. This opens subprocesses in a new
linux 'session' which starts a new process group with the pgid the same
as the trainers pid. Then when we send signals, we deliver them to the
process group rather than just the direct child.

[ghstack-poisoned]
